File: C++/HermitNF.cpp
// uses NTL
//   http://www.shoup.net/ntl

#include<NTL/mat_ZZ.h>
using namespace NTL;

void HermitNF(mat_ZZ& W, const mat_ZZ& A)
// A = matrix of shape (m,n) (m>=n)
// W = Hermite Normal Form A, i.e.,
//     lower triangular matrix of shape (n,n)
// reference: H. Cohen
//   "A Course in Computational Algebraic Number Theory"
//    Algorithm 2.4.5
{
    int m(A.NumRows()),n(A.NumCols());
    int i,j,k(m-1),l;
    ZZ d,s,t,u,v,b;
    if(m<n) Error("m<n in HermitNF");
    if(&W!=&A) W=A;
    for(l=n-1; l>=0; l--) {
        for(i=k-1; i>=0; i--) {
            if(IsZero(W[i][l])) continue;
            // DO NOT interchange
            // (u, W[i][l]) and (v, W[k][l]) in XGCD
            XGCD(d,u,v, W[i][l], W[k][l]);
            div(s, W[k][l], d);
            div(t, W[i][l], d);
            W[k][l] = d;
            clear(W[i][l]);
            for(j=0; j<l; j++) {
                mul(b, u, W[i][j]);
                MulAddTo(b, v, W[k][j]);
                W[i][j] *= s;
                MulSubFrom(W[i][j], t, W[k][j]);
                W[k][j] = b;
            }
        }
        if(IsZero(W[k][l])) continue;
        if(sign(W[k][l]) < 0)
           for(j=0; j<=l; j++)
               negate(W[k][j], W[k][j]);
        for(i=k+1; i<m; i++) {
            DivRem(s, W[i][l], W[i][l], W[k][l]);
            for(j=0; j<l; j++)
                MulSubFrom(W[i][j], s, W[k][j]);
        }
        k--;
    }
    if((k=m-n)==0) return;
    for(i=0; i<n; i++, k++)
        for(j=0; j<=i; j++)
            W[i][j] = W[k][j];
    W.SetDims(n,n);
}

File: HermitNF.py
from numpy import array
from ZZlib import XGCD

def HermitNF(A):
    """ Hermite Normal Form
    A = array of integers in shape(m,n)
    return lower triangular matrix in shape(n,n)
    referece: H. Cohen
      "A Course in Computational Algebraic Number Theory"
       Algorithm 2.4.5
    """
    A = array(A, dtype='object')
    if A.shape[0] < A.shape[1]:
        raise RuntimeError("wrong shape in HermitNF")
    k = A.shape[0]-1
    for l in range(A.shape[1]-1, -1, -1):
        for i in range(k-1,-1,-1):
            if A[i,l] == 0: continue
            d,u,v = XGCD(A[i,l], A[k,l])
            s,t = A[k,l]//d, A[i,l]//d
            A[k,l],A[i,l] = d,0
            A[i,:l],A[k,:l] = s*A[i,:l] - t*A[k,:l],\
                              u*A[i,:l] + v*A[k,:l]
        if A[k,l] == 0: continue
        if A[k,l] < 0: A[k,:l+1] = -A[k,:l+1]
        for i in range(k+1, A.shape[0]):
            q, A[i,l] = divmod(A[i,l], A[k,l])
            A[i,:l] -= q*A[k,:l]
        k -= 1

    return A[-A.shape[1]:]
